### ISSUE TITLE
Why is Adafruit MAX7219 Driver for 7 Segment Displays limited to Numeric Digits Only?

### ISSUE THREAD

**Issue by Jibun-no-Kage** [adafruit/Adafruit_CircuitPython_MAX7219#46]
Why is Adafruit MAX7219 Driver for 7 Segment Displays limited to Numeric Digits Only?  Well, dots and dashes are supported but no alpha characters at all?  This is kinda short sighted no?  Can you not please enhance this driver?  Would you please enhance this driver to include alpha characters?

**Comment by dotpointer** [comment]
I agree - this driver needs to be improved to allow more characters on 7 segment displays.

This is very possible on Raspberry Pi with regular Python, here is an example:
https://github.com/rm-hull/luma.led_matrix/blob/main/examples/sevensegment_demo.py

In the current state of this driver you can reach a few more letters using the set_digit(Position, Number) function, where Number can be 0-9, 10 = -, 11 = E, 12 = H, 13 = L, 14 = P, 15 = 0. The reason for these letters appears to be related to the simple test example, where the text "-HELP-90" is written:
http://www.circuitpython.cn/projects/max7219/en/latest/examples.html

Please enhance this driver to allow more than digits, these displays can be used for much more than numbers.

**Comment by ladyada** [comment]
to be honest we dont sell any displays with this chip - it was a contributed library. folks are encouraged to take over support or submit a PR!

**Comment by Jibun-no-Kage** [comment]
Just a suggestion... would this not be an opportunity to generate good will towards the community?  And support this solution given it is a popular solution?

**Comment by ladyada** [comment]
we agree!  please do the work and we will merge it in!

**Comment by Jibun-no-Kage** [comment]
LOL... I was suggesting that 'Adafruit' do the work.  But the comment put a smile on my face.

**Comment by dotpointer** [comment]
ADAFRUIT is 8 characters, it would fit nice in these 7 segment 8 digit displays. They are also cheap and could be a profitable product  in your inventory. Well, dream on the help is not coming so I made some investigation.

We start in the example:

examples/max7219_showbcddigits.py

```
value = 12345678
leds.show_str(0, f"{value:8}")
leds.show()
```

That show_str is here:

adafruit_max7219/bcddigits.py

`def show_str(self, start: int, strg: str) -> None:`

def show_str, replaces all characters not matching with blank, if we remove that then it is passed on to set_digit - or we use set_digit alone that is how HELP is written in the example:

`def set_digit(self, dpos: int, value: int) -> None:`

makes 4 loops with: 

`self.pixel(dpos, i, value & 0x01)`

where dpos = character position, i is the loop 0-3, and value is the character by number 0-15

The & value compares bits and only puts 1:s where they are equal (that is the explanation for the & operator as I have understood it)

adafruit_max7219/max7219.py

```
def pixel(self, xpos: int, ypos: int, bit_value: int = None) -> None:

Set one buffer bit

        :param int xpos: x position to set bit
        :param int ypos: y position to set bit
        :param int bit_value: value > 0 sets the buffer bit, else clears the buffer bit

bit_value = 0x01 if bit_value else 0x00
self.framebuf.pixel(xpos, ypos, bit_value)
```

So this passed on xpos = 0-7 that is the digit position, ypos = 0-3 from the loop for each digit and either  0x00 or 0x01 for bit_value if it is false or true

Adafruit_CircuitPython_framebuf/adafruit_framebuf.py: <-- note we have now switched repository

`def __init__(self, buf, width, height, buf_format=MVLSB, stride=None):`

No buf_format is supplied when init is called from the MAX7219 class, so we get the default, that is MVLSBFormat

```
class FrameBuffer:

def pixel(self, x, y, color=None):
```

A bunch of rotation handlings but it looks like it inits to 0 so we bypass that and reach:

```
if color is None:
            return self.format.get_pixel(self, x, y)
        self.format.set_pixel(self, x, y, color)
```

If color that is the character, return what is already there otherwise set it

And now we are at

```
class MVLSBFormat:

 @staticmethod
    def set_pixel(framebuf, x, y, color):
        """Set a given pixel to a color."""
        index = (y >> 3) * framebuf.stride + x
        offset = y & 0x07
        framebuf.buf[index] = (framebuf.buf[index] & ~(0x01 << offset)) | ((color != 0) << offset)

    @staticmethod
    def get_pixel(framebuf, x, y):
        """Get the color of a given pixel"""
        index = (y >> 3) * framebuf.stride + x
        offset = y & 0x07
        return (framebuf.buf[index] >> offset) & 0x01
```

Then we have show in the example, but that seems to forward the data to the SPI device.

We also have some examples for Micropython that has character lists and uses these to display more characters:

https://github.com/JennaSys/micropython-max7219

https://github.com/pdwerryhouse/max7219_8digit/blob/master/max7219_8digit.py

So we need to figure out how all that binary hex stuff above works and modify the library, or libraries if framebuf also needs charges.

**Comment by dotpointer** [comment]
@Jibun-no-Kage

I made some progress.

I have ported [micropython-max7219](https://github.com/JennaSys/micropython-max7219) to CircuitPython and it is now available at [circuitpython-max7219](https://github.com/dotpointer/circuitpython-max7219). It can write texts with letters, symbols and digits and it can display texts standing or scrolling and set brightness.

It is better than nothing - now seven segment displays can show texts running on ESP32:s with CircuitPython but it does not close this issue as this library should be updated to allow non-numeric texts.

**Comment by Jibun-no-Kage** [comment]
Yes, this is good step forward.  Well done.
